# U.S. Medical Insurance Costs

In this Porject, we are given a csv file **insurance.csv** that contains the dataset we will be working on for our analysis. The first step is to import the data from csv into our notebook file (ipnyb). <br> 

*See code below for importing, reading and iterating through the csv file to store values into variables.*

In [43]:
#To import csv library since we are working with a csv file
import csv

#Create empty lists to store values of each type of data
age = []
sex = []
bmi = []
children = []
smoker = []
region = []
charges = []
insurance_data = []

#Open file using DictReader, and iterate through each column to store value
#to their designated lists
with open('insurance.csv') as insurance_file:
    insurance_dict = csv.DictReader(insurance_file)
    for row in insurance_dict:
        age.append(row['age'])
        sex.append(row['sex'])
        bmi.append(row['bmi'])
        children.append(row['children'])
        smoker.append(row['smoker'])
        region.append(row['region'])
        charges.append(row['charges'])

#printing to check if values are stored as expected        
#print(age,sex,bmi,children,smoker,region,charges)     

#### Now that we have stored the values of our data into their respective lists, such as:

* age
* sex
* bmi
* children
* smoker
* region
* charges

We can now proceed to our analysis but first, we should create functions to get the values that we want such as the *MIN, MAX, AVG, MEAN, MEDIAN* ,etc.

#### Here are some data we might want to analyze:

* Organize 0 children, 1-3 children (count)
* Organize age brackets (under 18, 20s, 30s, 40s, 50s, 60s, and 65 year-olds)
* Organize male and females (count)
* average, median, and mean costs for both male and female
* Highest Cost (and details)
* Lowest Cost (and details)
* Count of each age (1-100)
* Average age of patients ()
* Median age of patients ()
* Where majority of individuals are from ()
* Cost avg between smokers and non-smokers ()
* Average charge of someone who has at least one child in this dataset ()


##### Organize 0 children, 1-3 children (count)

In [25]:
#Use count method to count number of people who has 0, 1, 2, 3, 4 and more children #list_name.count('value')
def count_children(children):
    count_0 = []
    count_1 = []
    count_2 = []
    count_3 = []
    count_4 = []
    count_5 = []
    count_6 = []
    for num_of_children in children:
        if num_of_children == '0':
            count_0.append(num_of_children)
        elif num_of_children == '1':
            count_1.append(num_of_children)
        elif num_of_children == '2':
            count_2.append(num_of_children)    
        elif num_of_children == '3':
            count_3.append(num_of_children)    
        elif num_of_children == '4':
            count_4.append(num_of_children)    
        elif num_of_children == '5':
            count_5.append(num_of_children)
        elif num_of_children == '6':
            count_6.append(num_of_children)    
    print("Number of people with 0 children: \t", count_0.count("0"), "\nNumber of people with 1 child: \t\t", count_1.count("1"),"\nNumber of people with 2 children: \t", count_2.count("2"), "\nNumber of people with 3 children: \t", count_3.count("3"), "\nNumber of people with 4 children: \t", count_4.count("4"), "\nNumber of people with 5 children: \t", count_5.count("5"), "\nNumber of people with 6 children: \t", count_6.count("6"))

(count_children(children))

Number of people with 0 children: 	 574 
Number of people with 1 child: 		 324 
Number of people with 2 children: 	 240 
Number of people with 3 children: 	 157 
Number of people with 4 children: 	 25 
Number of people with 5 children: 	 18 
Number of people with 6 children: 	 0


##### Organize age brackets (under 18, 20s, 30s, 40s, 50s, 60s, and 65 year-olds)

In [26]:
#Create a function to organize age brackets
def age_brackets(age):
    converted_age_list = [int(n) for n in age]
    under_18 = []
    twenties = []
    thirties = []
    fourties = []
    fifties = []
    sixties = []
    seniors = []
    for num in converted_age_list:
        if num <=18:
            under_18.append(num)
        elif num >=20 and num <30:
            twenties.append(num)
        elif num >=30 and num <40:
            thirties.append(num)
        elif num >=40 and num <50:
            fourties.append(num)
        elif num >=50 and num <60:
            fifties.append(num)
        elif num >=60 and num <65:
            sixties.append(num)
        elif num >=65 and num <100:
            seniors.append(num)
    print("No. of people who are underage: \t", len(under_18), "\nNo. of people who are in their 20s: \t", len(twenties), "\nNo. of people who are in their 30s: \t", len(thirties), "\nNo. of people who are in their 40s: \t", len(fourties), "\nNo. of people who are in their 50s: \t", len(fifties), "\nNo. of people who are over 60s: \t", len(sixties), "\nNo. of people who are seniors(65+): \t", len(seniors))
    
age_brackets(age)            

No. of people who are underage: 	 69 
No. of people who are in their 20s: 	 280 
No. of people who are in their 30s: 	 257 
No. of people who are in their 40s: 	 279 
No. of people who are in their 50s: 	 271 
No. of people who are over 60s: 	 114 
No. of people who are seniors(65+): 	 0


##### Organize male and females (count)

In [27]:
#Create a function to count number of males and females:
def count_sex(sex):
    male = []
    female = []
    for gender in sex:
        if gender == 'male':
            male.append(gender)
        else:
            female.append(gender)
    print("Total number of females: \t", female.count('female'), "\nTotal number of males: \t\t", male.count('male'))

count_sex(sex)    

Total number of females: 	 662 
Total number of males: 		 676


##### \*\*Average, Median, and Mean costs for all genders: ```Haven't learned mean, median yet using Python

In [28]:
converted_charges = [float(cost) for cost in charges]

total_cost_all = 0
for cost in converted_charges:
    total_cost_all += cost

rounded_total = round((total_cost_all), 2)    
print("\nThe total sum of all charges in the dataset is: $", rounded_total)
print("Total number of people: ", len(charges))
print("The average charge/cost in the dataset is: $", round(rounded_total/len(charges), 2))


The total sum of all charges in the dataset is: $ 17755824.99
Total number of people:  1338
The average charge/cost in the dataset is: $ 13270.42


##### \*\*Average, Median, and Mean costs for males:```Haven't learned mean, median yet using Python

In [107]:
#Create a list for the charges for all males
#first, create a list of tuples with sex and charges

sex_charges_list = []
n=0
for data in sex:
    sex_charges_list.append(data)
    sex_charges_list.append(charges[n])
    n+=1

#print(sex_charges_list)    
#print("\nLength: \t", len(sex_charges_list))    

n=0
total_cost_male = 0
for data in sex_charges_list:
    if data == 'male':
        total_cost_male += float(sex_charges_list[n+1])
        n+=1
    else:
        n+=1
        continue
        
print("Total sum of all charges for males: \t$", round((total_cost_male), 2)) 
print("The total average charge for all males is: $", round((round((total_cost_male), 2)/676),2))

Total sum of all charges for males: 	$ 9434763.8
The total average charge for all males is: $ 13956.75


##### \*\*Average, Median, and Mean costs for females:```Haven't learned mean, median yet using Python

In [30]:
#create a list of sum total charges for females
total_cost_female = 0
n=0
for data in sex_charges_list:
    if data == 'female':
        total_cost_female += float(sex_charges_list[n+1])
        n+=1
    else:
        n+=1

print("\nTotal sum of all charges for females: \t$", round((total_cost_female), 2))   
print("The average charges for all female in the dataset is: $", round((round((total_cost_female), 2)/662),2))


Total sum of all charges for females: 	$ 8321061.19
The average charges for all female in the dataset is: $ 12569.58


### Record with Highest Cost (and details):
###### First, we need to create a dictionary to get the details

In [31]:
#Creating the Dictionary
dataset_values = zip(age, sex, bmi, children, smoker, region, charges)
dataset_keys = range(1338)
dataset_dict = {key: value for key, value in zip(dataset_keys,dataset_values)}

#print(dataset_dict)  

def highest_cost(list):
    highest_charge = 0
    charges_list = [float(n) for n in list]
    for charge in charges_list:
        if charge >= highest_charge:
            highest_charge = charge
    return highest_charge

print("\nHighest charge is: $", highest_cost(charges))


Highest charge is: $ 63770.42801


In [32]:
#Check details (all other items) of the (value) in the dictionary using get method:

for record in dataset_dict:
    if dataset_dict[record][-1] == '63770.42801':
        print("Details of the person with the highest charges: ", dataset_dict.get(record))
        

Details of the person with the highest charges:  ('54', 'female', '47.41', '0', 'yes', 'southeast', '63770.42801')


##### Then, determine the record with Lowest Charge: 

In [33]:
def lowest_cost(list):
    lowest_charge = float('inf')
    charges_list = [float(n) for n in list]
    for charge in charges_list:
        if charge <= lowest_charge:
            lowest_charge = charge
    return lowest_charge

print("\nThe lowest charge is: $", lowest_cost(charges))


The lowest charge is: $ 1121.8739


In [34]:
#Check details (all other items) of the (value) in the dictionary using get method:

for record in dataset_dict:
    if dataset_dict[record][-1] == '1121.8739':
        print("Details of the person with the lowest charges: ", dataset_dict.get(record))

Details of the person with the lowest charges:  ('18', 'male', '23.21', '0', 'no', 'southeast', '1121.8739')


##### Count of each age in the record from 1-100: 

\#dataset_values = zip(age, sex, bmi, children, smoker, region, charges) <br>
\#dataset_dict = {range(1338): dataset_values}



In [35]:
# Dsclaimer: just copied this code from gpt, lul, would be better to use pandas i guess

# Creating variables with empty values using a for loop
for i in range(1, 101):
    globals()[f"age_{i}"] = None  # Assigning None as the empty value

In [36]:
#printing the variable names
#for name, value in locals().items():
    #print(name)

__name__
__doc__
__package__
__loader__
__spec__
__builtin__
__builtins__
_ih
_oh
_dh
In
Out
get_ipython
exit
quit
open
_
__
___
_i
_ii
_iii
_i1
csv
age
sex
bmi
children
smoker
region
charges
insurance_data
insurance_file
insurance_dict
row
_i2
count_children
_i3
age_brackets
_i4
count_sex
_i5
converted_charges
total_cost_all
cost
rounded_total
_i6
sex_charges_list
n
data
total_cost_male
_i7
total_cost_female
_i8
dataset_values
dataset_keys
dataset_dict
highest_cost
_i9
record
_i10
lowest_cost
_i11
_i12
i
age_1
age_2
age_3
age_4
age_5
age_6
age_7
age_8
age_9
age_10
age_11
age_12
age_13
age_14
age_15
age_16
age_17
age_18
age_19
age_20
age_21
age_22
age_23
age_24
age_25
age_26
age_27
age_28
age_29
age_30
age_31
age_32
age_33
age_34
age_35
age_36
age_37
age_38
age_39
age_40
age_41
age_42
age_43
age_44
age_45
age_46
age_47
age_48
age_49
age_50
age_51
age_52
age_53
age_54
age_55
age_56
age_57
age_58
age_59
age_60
age_61
age_62
age_63
age_64
age_65
age_66
age_67
age_68
age_69
age_70
age_71
a

In [42]:
def store_age(ages):
    age_counts = {}  # Dictionary to store distinct ages and their counts
    
    ages.sort() #sort it first before the loop
    
    # Iterate over the ages list
    for age in ages:
        if age in age_counts:
            age_counts[age] += 1  # Increment count if age already exists
        else:
            age_counts[age] = 1  # Initialize count if age is encountered for the first time

    return age_counts

age_count = store_age(age)

print(age_count)

{'18': 69, '19': 68, '20': 29, '21': 28, '22': 28, '23': 28, '24': 28, '25': 28, '26': 28, '27': 28, '28': 28, '29': 27, '30': 27, '31': 27, '32': 26, '33': 26, '34': 26, '35': 25, '36': 25, '37': 25, '38': 25, '39': 25, '40': 27, '41': 27, '42': 27, '43': 27, '44': 27, '45': 29, '46': 29, '47': 29, '48': 29, '49': 28, '50': 29, '51': 29, '52': 29, '53': 28, '54': 28, '55': 26, '56': 26, '57': 26, '58': 25, '59': 25, '60': 23, '61': 23, '62': 23, '63': 23, '64': 22}


In [38]:
#Print in formatted output:
for age, count in age_count.items():
    print("Age: {}, Count: {}".format(age,count))

# Anyone knows why the variable age changed value?    
print("Age list: ", age)    

Age: 18, Count: 69
Age: 19, Count: 68
Age: 20, Count: 29
Age: 21, Count: 28
Age: 22, Count: 28
Age: 23, Count: 28
Age: 24, Count: 28
Age: 25, Count: 28
Age: 26, Count: 28
Age: 27, Count: 28
Age: 28, Count: 28
Age: 29, Count: 27
Age: 30, Count: 27
Age: 31, Count: 27
Age: 32, Count: 26
Age: 33, Count: 26
Age: 34, Count: 26
Age: 35, Count: 25
Age: 36, Count: 25
Age: 37, Count: 25
Age: 38, Count: 25
Age: 39, Count: 25
Age: 40, Count: 27
Age: 41, Count: 27
Age: 42, Count: 27
Age: 43, Count: 27
Age: 44, Count: 27
Age: 45, Count: 29
Age: 46, Count: 29
Age: 47, Count: 29
Age: 48, Count: 29
Age: 49, Count: 28
Age: 50, Count: 29
Age: 51, Count: 29
Age: 52, Count: 29
Age: 53, Count: 28
Age: 54, Count: 28
Age: 55, Count: 26
Age: 56, Count: 26
Age: 57, Count: 26
Age: 58, Count: 25
Age: 59, Count: 25
Age: 60, Count: 23
Age: 61, Count: 23
Age: 62, Count: 23
Age: 63, Count: 23
Age: 64, Count: 22
Age list:  64


##### Average age of all patients

\#create a function that calculates the average age of all patients.<br>
Step 1. Create a variable to store total value
Step 2. get the average = total amount/total count

In [18]:
#I need to reread the file because somehow, at this point of the code, my list age changed the value to 64 and idk why

new_ages_list = []
with open('insurance.csv') as insurance_file:
    insurance_dict = csv.DictReader(insurance_file)
    for row in insurance_dict:
        new_ages_list.append(row['age'])

#To convert the age from 'str' to int:
age_list = [int(num) for num in new_ages_list]

#Create a function to calculate and return average by iterating through each age, adding all the values, and dividing by the length
def get_average_age(ages):
    
    #variable to store total amount
    total_amount = 0
    
    #iteration
    for num in age_list:
        total_amount+=num
    
    #Actual formula
    average_age = total_amount/len(age_list)
    
    #Just a clear formatting to output the results
    print("The total amount of ages is {} divided by {} number of people who bear the ages.".format(total_amount,len(age_list)))
    return round(average_age,2)    

print("The average age for the dataset is: {}".format(get_average_age(age_list)))

The total amount of ages is 52459 divided by 1338 number of people who bear the ages.
The average age for the dataset is: 39.21


##### Median age of patients (average of distinct ages):

In [52]:
distinct_ages = [int(key) for key in age_count]
#print(distinct_ages)

def median_age(unique_ages):
    
    total_distinct_ages = 0
    
    for age in unique_ages:
        total_distinct_ages += age
    
    return total_distinct_ages/len(unique_ages)

print("\nThe median age for the dataset is {}.".format(median_age(distinct_ages)))


The median age for the dataset is 41.0.


##### Count the region to count, there are 4 regions

In [58]:
#Function to count amount for each region
def count_region(regions):
    region_counts = {}  # Dictionary to store distinct regions and their counts
    
    regions.sort() #sort it before the loop
    
    # Iterate over the ages list
    for region in regions:
        if region in region_counts:
            region_counts[region] += 1  # Increment count if region already exists
        else:
            region_counts[region] = 1  # Initialize count if region is encountered for the first time
    
    return region_counts

counts_per_region = count_region(region)

print(counts_per_region)

{'northeast': 324, 'northwest': 325, 'southeast': 364, 'southwest': 325}


In [59]:
for key, value in counts_per_region.items():
    print("There are {} people from the {} region in the dataset.".format(value,key))

There are 324 people from the northeast region in the dataset.
There are 325 people from the northwest region in the dataset.
There are 364 people from the southeast region in the dataset.
There are 325 people from the southwest region in the dataset.


##### Cost avg between smokers and non-smokers


In [100]:
#print(dataset_dict)

#Create a function to store charges of smokers and non-smokers
def average_smoker_nonsmoker(dataset):
    
    smokers_charges = 0 #variable to store total charges for smokers
    nonsmoker_charges = 0 #variable to store total charges for non-smokers
    
    smokers_count = 0
    nonsmokers_count = 0
    
    for key, value in dataset.items():
        if value[4] == 'yes':
            smokers_charges += float(value[6])
            smokers_count += 1
        elif value[4] == 'no':
            nonsmoker_charges += float(value[6])   
            nonsmokers_count += 1
    
    smoker_avg = smokers_charges/smokers_count
    nonsmoker_avg = nonsmoker_charges/nonsmokers_count
    
    print("The average charge for a smoker is $ {}.\nThe average charge for a non-smoker is {}".format(round(smoker_avg,2),round(nonsmoker_avg,2)))

average_smoker_nonsmoker(dataset_dict)



The average charge for a smoker is $ 32050.23.
The average charge for a non-smoker is 8434.27


##### Average charge of someone who has at least one child in this dataset

In [106]:
def average_has_children(dataset):
    
    has_children_charges = 0
    no_children = 0
    
    has_children_count = 0
    no_children_count = 0
    
    for key, value in dataset.items():
        if int(value[3]) > 0:
            has_children_charges += float(value[6])
            has_children_count += 1
        elif int(value[3]) == 0:
            no_children += float(value[6])
            no_children_count += 1
    
    avg_has_children = has_children_charges/has_children_count
    avg_no_children = no_children/no_children_count
    
    print("The average charge for a parent is $ {}.\nThe average charge for a single is $ {}.".format(round(avg_has_children,2), round(avg_no_children,2))) 

average_has_children(dataset_dict)    

The average charge for a parent is $ 13949.94.
The average charge for a single is $ 12365.98.


# Current Findings for the dataset:

## Number of people who has children (0-6)
Number of people with 0 children: 	 574 <br>
Number of people with 1 child: 		 324 <br>
Number of people with 2 children: 	 240 <br>
Number of people with 3 children: 	 157 <br>
Number of people with 4 children: 	 25 <br>
Number of people with 5 children: 	 18 <br>
Number of people with 6 children: 	 0<br>

## Age Brackets (18-, 20s, 30s, 40s, 50s, 60s, 65+)
No. of people who are underage: 	 69 <br>
No. of people who are in their 20s: 	 280 <br>
No. of people who are in their 30s: 	 257 <br>
No. of people who are in their 40s: 	 279 <br>
No. of people who are in their 50s: 	 271 <br>
No. of people who are over 60s: 	 114 <br>
No. of people who are seniors(65+): 	 0<br>

## Total for each sex
Total number of females: 	 662 <br>
Total number of males: 		 676<br>

## Total Number of people, avg for males and females

Total number of people:  1338

The average charge/cost in the dataset is: $ 13270.42

The total average charge for all males is: $ 13956.75

The average charges for all female in the dataset is: $ 12569.58


## Min and Max Charges

Highest charge is: $ 63770.42801

Details of the person with the highest charges:  ('54', 'female', '47.41', '0', 'yes', 'southeast', '63770.42801') 

The lowest charge is: $ 1121.8739

Details of the person with the lowest charges:  ('18', 'male', '23.21', '0', 'no', 'southeast', '1121.8739')

## Parents and Singles (0 children vs has children):
The average charge for a parent is $ 13949.94.

The average charge for a single is $ 12365.98.

## Region

There are 324 people from the northeast region in the dataset. 

There are 325 people from the northwest region in the dataset. 

There are 364 people from the southeast region in the dataset. 

There are 325 people from the southwest region in the dataset.  

## Smoker 

The average charge for a smoker is $ 32050.23.

The average charge for a non-smoker is 8434.27 

## Age

The average age for the dataset is: 39.21

The median age for the dataset is 41.0.

##### Ages and count
Age: 18, Count: 69 <br>
Age: 19, Count: 68 <br>
Age: 20, Count: 29 <br>
Age: 21, Count: 28 <br>
Age: 22, Count: 28 <br>
Age: 23, Count: 28 <br>
Age: 24, Count: 28 <br>
Age: 25, Count: 28 <br>
Age: 26, Count: 28 <br>
Age: 27, Count: 28 <br>
Age: 28, Count: 28 <br>
Age: 29, Count: 27 <br>
Age: 30, Count: 27 <br>
Age: 31, Count: 27 <br>
Age: 32, Count: 26 <br>
Age: 33, Count: 26 <br>
Age: 34, Count: 26 <br>
Age: 35, Count: 25 <br>
Age: 36, Count: 25 <br>
Age: 37, Count: 25 <br>
Age: 38, Count: 25 <br>
Age: 39, Count: 25 <br>
Age: 40, Count: 27 <br>
Age: 41, Count: 27 <br>
Age: 42, Count: 27 <br>
Age: 43, Count: 27 <br>
Age: 44, Count: 27 <br>
Age: 45, Count: 29 <br>
Age: 46, Count: 29 <br>
Age: 47, Count: 29 <br>
Age: 48, Count: 29 <br>
Age: 49, Count: 28 <br>
Age: 50, Count: 29 <br>
Age: 51, Count: 29 <br>
Age: 52, Count: 29 <br>
Age: 53, Count: 28 <br>
Age: 54, Count: 28 <br>
Age: 55, Count: 26 <br>
Age: 56, Count: 26 <br>
Age: 57, Count: 26 <br>
Age: 58, Count: 25 <br>
Age: 59, Count: 25 <br>
Age: 60, Count: 23 <br>
Age: 61, Count: 23 <br>
Age: 62, Count: 23 <br>
Age: 63, Count: 23 <br>
Age: 64, Count: 22 <br>